text/3663-feature-deps.md

```diff
@@ -0,0 +1,112 @@
+this microformat adds complexity compared to other parts of the Cargo manifest,
+which distinguish different kinds of things via different keys or sections. (By
+way of example, we have `dev-dependencies` and `build-dependencies`, rather
+than a microformat like `dependencies."build:xyz"`.)
+
+Now that [RFC 3416](https://github.com/rust-lang/rfcs/pull/3416) has defined a
+structured table form for defining features, this RFC takes a first step
+towards providing structured equivalents in place of the microformat.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When defining features in a Cargo manifest, you can specify a feature's
+dependencies in a separate key `deps`. You can define a feature that depends on
+both other features and on optional crates, or on only one or the other.
+
+```toml
+[features]
+feat1 = { enables = ["feat2", "feat3"], deps = ["crate1"] }
+feat2 = { deps = ["crate2"] }
+feat3.deps = ["crate3"]
+"""
+```
+
+The `deps` key is equivalent to including items in `enables` (or a traditional
+feature definition) with a `dep:` prefix. For example,
+`feat2 = { deps = ["crate2"] }` is equivalent to
+`feat2 = { enables = ["dep:crate2"] }` or
+`feat2 = ["dep:crate2"]`.
+
+However, if a crate uses the `deps` key, it should consistently use the `deps`
+key for all feature dependencies on crates, rather than using a mix of `deps`
+and `"dep:"`. (Cargo may choose to warn about mixing the two.)
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The index can translate this to the microformat, to preserve compatibility
+until the next time the index format needs to make an incompatible change.
+
+Cargo should detect invalid syntax in the `deps` key, such as the use of `dep:`
+or `cratename/featurename` or `cratename?/featurename`, and provide errors that
+suggest what to write instead.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- This adds minor incremental complexity to parsing the manifest, both for
+  Cargo itelf, and for third-party tools that parse the manifest without using
+  Cargo's parser.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+We could translate this to the microformat when creating the `Cargo.toml` on
+publish, but this would not work for features using any new fields (such as
+`doc`), so this RFC proposes that using the new mechanism require new Cargo,
+for simplicity.
+
+## Naming
+
+- `deps` is consistent with the existing `"dep:"` prefix, and is shorter so it
+  more easily fits in a one-line feature definition.
+- `dependencies` would be more consistent with the top-level `dependencies`
+  tables.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+In the future, we could also provide an alternative to the
+`crate-name?/feature-name` microformat, or alternatively we could migrate
+towards making `crate-name/feature-name` mean that in a future edition. This
```

Comment:
My -1 is mainly because of putting `pkg?/feat` and `pkg/feat` into "future possibilities" 😕. Without "structurizing" DepFeature, the introduction of `deps` key puts the transition in an awkward incomplete state as you still need to use the microformat to enable features of a dep. Better just postpone if it's not ready.

I don't think actually defining DepFeature here makes the RFC harder to review, this RFC is already short enough.

Reply:
> as you still need to use the microformat to enable features of a dep

Agree. Not considering `DepFeature` as a whole may end up having three ways to define that. It's better to one of the other, not a combination of them.

```toml
# old
[features]
tracing = ["valuable", "log/std"]

# this
[features]
tracing = { enables = ["valuable", "log/std"], deps = ["log"]  }

# future maybe?
[features.tracing]
enables = ["valuable"]
deps.log = ["std"]

```

Reply:
I agree from at least a specific perspective.  For me, the thing that has bothered me is the inconsistent way of referring to dependencies:
```
feature_name
dep:dep_name
dep_name/feature_name
```
I feel like this should be
```
feature_name
dep:dep_name
dep:dep_name/feature_name
```
(this is worse with implicit features because deps and features get muddled but 2024 fixes that)